streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java

```diff
@@ -184,6 +247,75 @@ public void process(final Record<KIn, VIn> record) {
                     droppedRecordsSensor.record();
                 }
             }
+
+            maybeMeasureLatency(() -> tryEmitFinalResult(record, closeTime), time, emitFinalLatencySensor);
+        }
+
+        private void tryEmitFinalResult(final Record<KIn, VIn> record, final long closeTime) {
+            if (emitStrategy.type() != StrategyType.ON_WINDOW_CLOSE) {
+                return;
+            }
+
+            final long now = internalProcessorContext.currentSystemTimeMs();
+            // Throttle emit frequency
+            if (now < timeTracker.nextTimeToEmit) {
+                return;
+            }
+
+            // Schedule next emit time based on now to avoid the case that if system time jumps a lot,
+            // this can be triggered everytime
+            timeTracker.nextTimeToEmit = now;
+            timeTracker.advanceNextTimeToEmit();
+
+            // Close time does not progress
+            if (lastEmitCloseTime != ConsumerRecord.NO_TIMESTAMP && lastEmitCloseTime >= closeTime) {
+                return;
+            }
+
+            final long emitWindowStart = closeTime - windows.size();
+            if (emitWindowStart < 0) {
+                // If emitWindowStart is 0, it means first window closes since windowEndTime
+                // is exclusive
+                return;
+            }
+
+            // Because we only get here when emitWindowStart > 0 which means closeTime > windows.size()
+            // Since we set lastEmitCloseTime to closeTime before storing to processor metadata
+            // lastEmitCloseTime - windows.size() is always > 0
+            // Set lastEmitWindowStart to -1L if not set so that when we fetchAll, we fetch from 0L
+            final long lastEmitWindowStart = lastEmitCloseTime == ConsumerRecord.NO_TIMESTAMP ?
+                -1L : lastEmitCloseTime - windows.size();
+
+            if (lastEmitCloseTime != ConsumerRecord.NO_TIMESTAMP) {
+                final Map<Long, W> matchedCloseWindows = windows.windowsFor(emitWindowStart);
+                final Map<Long, W> matchedEmitWindows = windows.windowsFor(lastEmitWindowStart);
+
+                // Don't fetch store if the new emit window close time doesn't progress enough to cover next
+                // window
+                if (matchedCloseWindows.equals(matchedEmitWindows)) {
```

Comment:
Is this the same as `emitWindowStart ==  lastEmitWindowStart`? If yes, comparing two longs seems to be more efficient than comparing to `Maps`

Reply:
As we `.fetchAll(lastEmitWindowStart + 1, emitWindowStart);` and both upper and lower bound are inclusive, would we not need that `emitWindowStart >= lastEmitWindowStart+1` to have at least one window that we could potentially emit? Otherwise, lower bound is larger than upper bound and the time-range is not well defined?

Reply:
> Is this the same as emitWindowStart == lastEmitWindowStart

Nope. They may not be the same. `emitWindowStart` is based on current `closeTime` and `lastEmitWindowStart` is based last emitted close time. These are not the same. The logic here to detect if `closeTime` doesn't progress much to emit new windows.

For example, for tumbling window length 10sec and close time 11, window [0, 10) will be emitted, if close time progress to 12, window to emit is still [0, 10).